components/raftstore/src/store/peer.rs

```diff
@@ -3696,16 +3704,72 @@ where
 
         let transfer_leader = get_transfer_leader_cmd(&req).unwrap();
         let peer = transfer_leader.get_peer();
-
-        let transferred = self.pre_transfer_leader(peer);
-
+        let transferred = if self.peer.id == peer.id {
+            true
+        } else {
+            match self.propose_locks_before_transfer_leader(ctx) {
+                Ok(true) => self.propose_normal(ctx, req).is_ok(),
+                Ok(false) => self.pre_transfer_leader(peer),
+                Err(_) => false,
+            }
+        };
         // transfer leader command doesn't need to replicate log and apply, so we
         // return immediately. Note that this command may fail, we can view it just as an advice
         cb.invoke_with_response(make_transfer_leader_response());
 
         transferred
     }
 
+    // Returns whether it needs to propose the transfer leader command instead of
+    // just sending messages.
+    fn propose_locks_before_transfer_leader<T>(
+        &mut self,
+        ctx: &mut PollContext<EK, ER, T>,
+    ) -> Result<bool> {
+        // 1. Disable in-memory pessimistic locks
+        let mut pessimistic_locks = self.txn_ext.pessimistic_locks.write();
+        let locks = if pessimistic_locks.valid {
+            // FIXME: `valid` should be set back to true if the peer remains being leader after
+            // some time (which means leader transfer fails)
+            pessimistic_locks.valid = false;
+            // If there is no lock when setting `valid` to false, we can guarantee that no lock
+            // needs to be migrated to the new leader. In this case, we can just send a message
+            // to trigger leader transfer instead of making a proposal.
+            if pessimistic_locks.map.is_empty() {
+                return Ok(false);
+            }
+            mem::take(&mut pessimistic_locks.map)
+        } else {
+            // If `valid` is false, the locks may have been proposed but not committed.
+            // So, it is not safe to use a Raft message to transfer leader in case the lock
```

Comment:
I think it's safe because the leader will wait for the transferee's log progress and all memory lock requests have been proposed to the leader.

Reply:
I'm thinking that Raft messages don't keep the order with proposals (potentially because of internal retry mechanisms). To my knowledge such a raft message can be prior to a previously proposed command...

Reply:
> To my knowledge such a raft message can be prior to a previously proposed command

Yes, but the leader has to replicate all raft logs to the transferee before it transfers leadership to the transferee.